docs/gethue/content/en/posts/2021-03-04-vue3-migration.md

```diff
@@ -0,0 +1,269 @@
+        },
+        created() {
+          this.$watch(
+            ():K => this.foo.bar, // Watch nested property bar of type K
+            (val:K, prevVal:K): void => {
+              // Statements to be executed
+            }
+          )
+        }
+      })
+    </script>
+
+## Web components
+
+Our next goal was packaging the component to be shared across various projects. As modern Web UIs are built using various technologies we needed a method that's framework agnostic. Enter the web component! Web Components allows you to create reusable custom elements with their functionality encapsulated away from the rest of your code.
+
+But to our surprise the official Vue package for converting a component to a web component did not support Vue 3. And as per [this issue](https://github.com/vuejs/vue-web-component-wrapper/issues/93) it's going to be a while before the support is added. So we had to find an alternative. We created a port of the Vue 2 web component wrapper that works with Vue 3. The code is [here](https://github.com/cloudera/hue/tree/master/desktop/core/src/desktop/js/vue/wrapper) and npm package is available [here](https://www.npmjs.com/package/vue3-webcomponent-wrapper). Our port supports reactive attributes, events & slots.
+
+_Note: One main blocker preventing the official wrapper for upgrading was the lack of shadow-root CSS injection in Vue 3 build tooling. As we could live without shadow in Hue this was not an issue and port took hardly a day._
+
+Using our component wrapper is pretty easy. It can be installed using `npm i --save vue3-webcomponent-wrapper`. Once installed following snippet show how to create a custom tag named `my-component`.
+
+    import { createApp, h } from "vue";
+    import wrapper from "vue3-webcomponent-wrapper";
+    import MyComponent from "./components/MyComponent.vue";
+
+    const webComponent = wrapper(MyComponent, createApp, h);
+    window.customElements.define("my-component", webComponent);
+
+Please find more information about the wrapper in this [demo app](https://github.com/sreenaths/vue3-webcomponent-wrapper-demo). And [this](https://github.com/cloudera/hue/blob/master/apps/metastore/src/metastore/templates/metastore.mako#L825) er-diagram tag is a sample use of the wrapper in Hue project.
+
+### Using official Vue 3 web component wrapper
+
+The official wrapper must be very similar to our implementation. That said, the exact function signature of the official wrapper is unknown at this point. Keep a check on [this page](https://www.npmjs.com/package/vue3-webcomponent-wrapper) for future updates.
+
+## And that's it!
+
+Hue moved from using pretty old technologies to the forefront of Web interface development with Vue 3 and the component wrapper. We strongly believe component programming is a very effective development paradigm with its isolation and easy sharing. In the next episode we will demo how it all integrates with the new Hue 5 API.
+
+~ Sreenath & Romain from the Hue Team
```

Comment:
No need to put my name, remove it ;)

Reply:
Without you the blog wouldn't be good enough 🙂